docs: gpbackup/gprestore updates

mkiyama opened the pull request:
Summary of updates for GA.

-Remove experimental notes (GA release)
-Update email configuration information
-Add return code information
-Fix syntax error in gpbackup

Link to HTML in GPDB Review doc site.
Admin. Guide
https://docs-gpdb-review-staging.cfapps.io/550/admin_guide/managing/backup-gpbackup.html

Utility Guide
http://docs-gpdb-review-staging.cfapps.io/550/utility_guide/admin_utilities/gpbackup.html
http://docs-gpdb-review-staging.cfapps.io/550/utility_guide/admin_utilities/gprestore.html


PR for 5X_STABLE 
Will be ported to MAIN

mkiyama commented:
@khuddlefish 
Question about email notification.
If the config file has status: success, does that mean return code 0 or 1?

thanks

dyozie commented on gpdb-doc/dita/utility_guide/admin_utilities/gprestore.xml:
```diff
@@ -52,6 +49,13 @@
         operations.</p>
       <p><codeph>gprestore</codeph> does not currently support filtering the database objects that
         are restored from a backup set. </p>
+      <p>When a rest9re operation completes, <codeph>gprestore</codeph> returns a a status code. See
```
Typo:  rest9re

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -202,22 +199,38 @@ $ <b>gprestore -timestamp 20171103152558 -createdb</b>
+        <title>Return Codes</title>
+        <p>One of these codes is returned after <codeph>gpbackup</codeph> or
+            <codeph>gprestore</codeph> completes.<ul id="ul_hhb_sqp_n4">
+            <li><b>0</b> – Backup or restore completed with no problems </li>
```
Add period.

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -339,17 +352,111 @@ public.sales_1_prt_dec17</codeblock>
+        a back up or restore operation completes. </p>
+      <p>To have <codeph>gpbackup</codeph> or <codeph>gprestore</codeph> send out status email
+        notifications, you must place a file named <codeph>gp_email_contacts.yaml</codeph> in the
+        home directory of the user running <codeph>gpbackup</codeph> or <codeph>gpbackup</codeph> in
```
change second gpbackup to gprestore.

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -339,17 +352,111 @@ public.sales_1_prt_dec17</codeblock>
+        notifications, you must place a file named <codeph>gp_email_contacts.yaml</codeph> in the
+        home directory of the user running <codeph>gpbackup</codeph> or <codeph>gpbackup</codeph> in
+        the same directory as the utilities (<codeph>$GPHOME/bin</codeph>). The utilities issue a
+        message if it cannot locate a <codeph>gp_email_contacts.yaml</codeph> file in either
```
"it" -> "they"

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -339,17 +352,111 @@ public.sales_1_prt_dec17</codeblock>
+        message if it cannot locate a <codeph>gp_email_contacts.yaml</codeph> file in either
+        location. If both locations contain a <codeph>.yaml</codeph> file, the utility uses the file
+        in user <codeph>$HOME</codeph>. </p>
+      <p>The email subject line include the utility name and timestamp and the name of the Greenplum
```
include -> includes

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -339,17 +352,111 @@ public.sales_1_prt_dec17</codeblock>
+              <pd>Optional. Begins the <codeph>gprestore</codeph> email section. This section
+                contains the <codeph><xref href="#topic_uq5_v5v_scb/address_yml" format="dita"
+                    >address</xref></codeph> and <codeph><xref href="#topic_uq5_v5v_scb/status_yml"
+                    format="dita">status</xref></codeph> parameters that are used to send email
```
"send email" -> "send an email"

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -378,26 +485,26 @@ public.sales_1_prt_dec17</codeblock>
-                    <codeph>-dbname)</codeph> that must be created <i>prior</i> to restoring the
-                  actual data. These objects include:<ul id="ul_s35_nhj_tbb">
+                  </ul><p><b>Note:</b> Global metadata is not restored by default. You must include
+                    the <codeph>-globals</codeph> options to <codeph>gprestore</codeph> to restore
```
options -> option

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -378,26 +485,26 @@ public.sales_1_prt_dec17</codeblock>
+                  </ul><p><b>Note:</b> Global metadata is not restored by default. You must include
+                    the <codeph>-globals</codeph> options to <codeph>gprestore</codeph> to restore
+                    global metadata.</p>Database-specific objects that must be created <i>before</i>
+                  to restoring the actual data include:<ul id="ul_vqn_vsv_scb">
```
remove "to"

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -412,13 +519,8 @@ public.sales_1_prt_dec17</codeblock>
-                <entry>Contains DDL for objects in the backed-up database (specified with
-                    <codeph>-dbname)</codeph> that must be created <i>after</i> restoring the data.
-                  These objects include:<ul id="ul_zyg_tgj_tbb">
+                  </ul>Database-specific objects that must be created <i>after</i> to restoring the
```
remove "to"

dyozie commented on gpdb-doc/dita/utility_guide/admin_utilities/gpbackup.xml:
```diff
@@ -44,19 +41,20 @@
+        all backup files from the Greenplum Database master and segment hosts to an absolute path
+        for later use. Additional options are provided to filter the backup set in order to include
+        or exclude specific tables.</p><p>Each <codeph>gpbackup</codeph> task uses a single
+        transaction on the Greenplum database master host. Utility-mode connections are created for
```
Capitalize: "Database"

dyozie commented on gpdb-doc/dita/utility_guide/admin_utilities/gpbackup.xml:
```diff
@@ -44,19 +41,20 @@
+        each segment host, which perform the associated <codeph>COPY ... ON SEGMENT</codeph>
+        operations in parallel. The backup process acquires an <codeph>ACCESS SHARE</codeph> lock on
+        each table that is backed up.</p>When a back up operation completes,
+        <codeph>gpbackup</codeph> returns a a status code. See <xref
```
repeated word:  "a a"

dyozie commented on gpdb-doc/dita/utility_guide/admin_utilities/gpbackup.xml:
```diff
@@ -260,6 +258,16 @@
+      <title>Return Codes</title>
+      <p>One of these codes is returned after <codeph>gpbackup</codeph> completes.<ul
+          id="ul_hhb_sqp_n4">
+          <li><b>0</b> – Backup completed with no problems </li>
```
Add period (.)

khuddlefish commented:
@mkiyama success will be status 0, success_with_errors will be status 1, failure will be status 2.

mkiyama commented:
commits for review edits and changes to option names
-backup-dir, -with-globals, -redirect-db

Review site HTML has also been updated.

khuddlefish commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -339,17 +353,149 @@ public.sales_1_prt_dec17</codeblock>
+                          <pt id="status_yml"><b>status</b></pt>
+                          <pd>Optional. Specify when the utility sends an email to the specified
+                            email address. The default is to send an email on either a successful
+                            (with or without errors) or failed operation.</pd>
```
The default information here is wrong. If status is not specified, no email will be sent.

mkiyama commented:
updated email notification section
-the default for the status section is to not send email notification.
-fixed the example to match section default.

mkiyama commented:
updated gprestore to state new database is cloned from template0.

Review site HTML has also been updated.

khuddlefish commented on gpdb-doc/dita/utility_guide/admin_utilities/gprestore.xml:
```diff
@@ -7,21 +7,18 @@
    [<b>-createdb</b>]
    [<b>-debug</b>]
-   [<b>-globals</b>]
+   [<b>-with-globals</b>]
```
Can we reorder the flags to be in alphabetical order?

khuddlefish commented on gpdb-doc/dita/utility_guide/admin_utilities/gprestore.xml:
```diff
@@ -191,24 +208,22 @@ $ gprestore -timestamp 20171103152558 -createdb</codeblock></p>
       <p>Restore the backup set to the "demos" database instead of the "demo" database that was
```
demo2 not demos

khuddlefish commented on gpdb-doc/dita/utility_guide/admin_utilities/gprestore.xml:
```diff
@@ -191,24 +208,22 @@ $ gprestore -timestamp 20171103152558 -createdb</codeblock></p>
+        connections:<codeblock>$ gprestore -backup-dir /home/gpadmin/backups/ -timestamp 20171103153156 -createdb -jobs 8</codeblock></p>
       <p>Restore only the "wikipedia" schema included in the backup
         set:<codeblock>$ dropdb demo
 $ gprestore -include-schema wikipedia /home/gpadmin/backups/ -timestamp 20171103153156 -createdb</codeblock></p>
```
I'm not sure what argument the "/home/gpadmin/backups/" path refers to.

khuddlefish commented on gpdb-doc/dita/utility_guide/admin_utilities/gprestore.xml:
```diff
@@ -31,27 +28,35 @@
         cluster. You can optionally restore a backup set to a different database by using the
-          <codeph>-redirect</codeph> option.</p>
+          <codeph>-redirect-db</codeph> option.</p>
       <p>Database objects to be restored must not exist in the target database. If the database
```
The line about using `-createdb` to create the database is redundant with the line above.

khuddlefish commented on gpdb-doc/dita/utility_guide/admin_utilities/gpbackup.xml:
```diff
@@ -36,27 +33,28 @@
+        all backup files from the Greenplum Database master and segment hosts to an absolute path
+        for later use. Additional options are provided to filter the backup set in order to include
+        or exclude specific tables.</p><p>Each <codeph>gpbackup</codeph> task uses a single
+        transaction on the Greenplum Database master host. Utility-mode connections are created for
```
We do not connect in utility mode for gpbackup. COPY ... ON SEGMENT is called from the master and the database handles the parallelism.

khuddlefish commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -339,17 +353,144 @@ public.sales_1_prt_dec17</codeblock>
+                    <pd><varname>user</varname>@<varname>domain</varname> is a single, valid email address.<parml>
+                        <plentry>
+                          <pt id="status_yml"><b>status</b></pt>
+                          <pd>Optional. Specify when the utility sends an email to the specified
```
It might make sense to say status is required since it doesn't make a lot of sense for them to specify an address in a file, but not want to get any emails. We also want to make sure that we are clear that each address gets its own status block, and you can specify multiple address/status sections together, not just multiple addresses. The current docs are a little ambiguous around that.

khuddlefish commented:
Not that it needs to be part of this same pull request, but there is also some information about filtering that is incorrect. I wasn't able to comment on it here. Additionally, it might be good to talk more about the single-data-file option. We barely mention it, but it does have some very important use cases.

mkiyama commented:
@khuddlefish 
let me know what the changes should be.
I can create a new PR, or generate a PDF if that would be easier.

mkiyama commented:
pushed commit with updates

Review site HTML has also been updated.

List of changes

AdminGuide
updated email notification informmation
updated gpbackup/gprestore output
added report file information
fixed partition table filtering example
corrected gpbackup schema/table filtering support
updated restore limitation. table index, rule, trigger are restored

gpbackup doc
change the gpbackup option name from -backupdir to -backup-dir
corrected description of parallel backup


gprestore doc
change the gprestore option name from -createdb to -create-db
updated description of restore operation in description
updated restore description if schema exists in database
updated restore limitation. table index, rule, trigger are restored

mkiyama commented:
@oakbarrett @khuddlefish 
Do you some text about -single-data-file use cases.
This was mentioned in an earlier comment by Karen

oakbarrett commented:
--single-data-file is recommended in situations where additional overhead of having multiple files may be prohibitive, as is the case with third party storage solutions such as Data Domain.

mkiyama commented:
@khuddlefish @jmcatamney @oakbarrett 

added new commit.
-new restore options
-clarified leaf partition backup/restore

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -183,6 +185,9 @@ $ <b>find /home/gpadmin/backups/ -type f</b>
 /home/gpadmin/backups/gpseg-1/backups/20171103/20171103153156/gpbackup_20171103153156_report
 /home/gpadmin/backups/gpseg-1/backups/20171103/20171103153156/gpbackup_20171103153156_toc.yaml</codeblock></p>
+      <p>When performing a backup operation, you can use the <codeph>--single-data-file</codeph> in
+        situations where the additional overhead multiple files might be prohibitive. For example,
```
overhead -> overhead of

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -183,6 +185,9 @@ $ <b>find /home/gpadmin/backups/ -type f</b>
 /home/gpadmin/backups/gpseg-1/backups/20171103/20171103153156/gpbackup_20171103153156_toc.yaml</codeblock></p>
+      <p>When performing a backup operation, you can use the <codeph>--single-data-file</codeph> in
+        situations where the additional overhead multiple files might be prohibitive. For example,
+        if you use a third party storage solutions such as Data Domain with back ups.</p>
```
solutions -> solution

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -364,11 +369,13 @@ public.sales_1_prt_dec17 </codeblock></p><p>Then
-          a file specified with <codeph>-exclude-table-file</codeph>, <codeph>gpbackup</codeph>
-          ignores the partition names.</note></section>
+          partition:<codeblock>$ <b>gpbackup -dbname demo -include-table public.sales -leaf-partition-data</b></codeblock><p>When
+          a leaf partitions are backed up, the leaf partition data is backed up along with the
```
a leaf -> leaf

dyozie commented on gpdb-doc/dita/admin_guide/managing/backup-gpbackup.xml:
```diff
@@ -364,11 +369,13 @@ public.sales_1_prt_dec17 </codeblock></p><p>Then
+          partition:<codeblock>$ <b>gpbackup -dbname demo -include-table public.sales -leaf-partition-data</b></codeblock><p>When
+          a leaf partitions are backed up, the leaf partition data is backed up along with the
+          metadata for the entire partitioned table.</p><note>You cannot use the
+            <codeph>-exclude-table-file</codeph> with <codeph>-leaf-partition-data</codeph>.
```
with -> option with

dyozie commented on gpdb-doc/dita/utility_guide/admin_utilities/gprestore.xml:
```diff
@@ -102,6 +112,45 @@
+            letter, number, or an underscore character, then you must include that name in double
+            quotes. If a table is not in the backup set, the restore operation fails. You cannot
+            specify a leaf partition of a partitioned table.</pd>
+          <pd>When restoring a leaf partition, the partitioned table is also restored. You can
```
"You can specify a leaf partition of a partitioned table only if the leaf partition was listed as a table to be backed up."

This conflicts with the sentence in --exclude-table:  "You cannot specify a leaf partition of a partitioned table."

dyozie commented:
Going to merge this PR now.  @oakbarrett - if any changes are needed to the last commits (new option descriptions) let's handle it as a fix to the docs.
